shared/js/background/atb.es6.js

```diff
@@ -126,24 +129,42 @@ const ATB = (() => {
             return atb
         },
 
+        // iterate over a list of accepted params, and retrieve them from a URL
+        // builds a new query string containing only accepted params
+        getAcceptedParamsFromURL: (url) => {
+            let validParams = new URLSearchParams()
+            const parsedUrl = new URL(url)
+
+            ACCEPTED_URL_PARAMS.forEach(param => {
+                if (parsedUrl.searchParams.has(param)) {
+                    validParams.append(param, parsedUrl.searchParams.get(param))
+                }
+            })
+
+            return validParams.toString()
+        },
+
         updateATBValues: () => {
             // wait until settings is ready to try and get atb from the page
             return settings.ready()
                 .then(ATB.setInitialVersions)
                 .then(browserWrapper.getDDGTabUrls)
                 .then((urls) => {
                     let atb
+                    let params
 
                     urls.some(url => {
                         atb = ATB.getNewATBFromURL(url)
+                        params = ATB.getAcceptedParamsFromURL(url)
                         return !!atb
                     })
 
                     if (atb) {
                         settings.updateSetting('atb', atb)
                     }
 
-                    ATB.finalizeATB()
+                    // only pass params if we have an ATB
+                    ATB.finalizeATB(atb ? params : '')
```

Comment:
This works perfectly, and very much :+1: on using the URL API in `getAcceptedParamsFromURL`.

I was wondering if we can generalize it further. E.g. we're reading the URL twice (in `getNewATBFromURL` and `getAcceptedParamsFromURL`) and sorta serializing it twice (in `getAcceptedParamsFromURL` and `finalizeATB`).

Maybe we could grab both `natb` and `cp` in `getAcceptedParamsFromURL`, do any validation within (e.g. correct ATB format, renaming from `natb` to `atb`) and return an object. Then maybe we can pass URL params wholesale to `finalizeATB`.

What do you think?

Reply:
I had almost implemented something like this, but opted to not change the current ATB logic to keep things simple. I totally agree tho that we're duplicating efforts with parsing the URL twice. I'll go ahead with your suggested approach :+1: